game-app/game-core/src/main/java/games/strategy/triplea/delegate/TechTracker.java

```diff
@@ -11,73 +12,204 @@
+    return getCached(player, type, "canBlitz", getter) == 1;
   }
 
   public boolean canBombard(GamePlayer player, UnitType type) {
-    return getUnitAbilitiesGained(TechAbilityAttachment.ABILITY_CAN_BOMBARD, type, player);
+    final Supplier<Integer> getter =
+        () -> getUnitAbilitiesGained(TechAbilityAttachment.ABILITY_CAN_BOMBARD, type, player);
+    return getCached(player, type, "canBombard", getter) == 1;
+  }
+
+  public int getMinimumTerritoryValueForProductionBonus(final GamePlayer player) {
+    final Supplier<Integer> getter =
+        () ->
+            Math.max(
+                0,
+                getCurrentTechAdvances(player).stream()
+                    .map(TechAbilityAttachment::get)
+                    .filter(Objects::nonNull)
+                    .mapToInt(TechAbilityAttachment::getMinimumTerritoryValueForProductionBonus)
+                    .filter(i -> i != -1)
+                    .min()
+                    .orElse(-1));
+    return getCached(player, null, "getMinimumTerritoryValueForProductionBonus", getter);
+  }
+
+  public int getRocketNumberPerTerritory(final GamePlayer player) {
+    final Supplier<Integer> getter =
+        () ->
+            sumNumbers(
+                TechAbilityAttachment::getRocketNumberPerTerritory,
+                TechAdvance.TECH_NAME_ROCKETS,
+                getCurrentTechAdvances(player));
+    return getCached(player, null, "getRocketNumberPerTerritory", getter);
+  }
+
+  public int getRocketDistance(final GamePlayer player) {
+    final Supplier<Integer> getter =
+        () ->
+            sumNumbers(
+                TechAbilityAttachment::getRocketDistance,
+                TechAdvance.TECH_NAME_ROCKETS,
+                getCurrentTechAdvances(player));
+    return getCached(player, null, "getRocketDistance", getter);
+  }
+
+  private int getCached(
+      GamePlayer player, UnitType type, String property, Supplier<Integer> getFunction) {
+    if (!ENABLE_CACHING) {
+      return getFunction.get();
+    }
+    return cache.computeIfAbsent(new Key(player, type, property), key -> getFunction.get());
   }
 
   private int getSumOfBonuses(
       Function<TechAbilityAttachment, IntegerMap<UnitType>> mapper,
       UnitType type,
       GamePlayer player) {
-    return TechAbilityAttachment.sumIntegerMap(mapper, type, getCurrentTechAdvances(player));
+    return sumIntegerMap(mapper, type, getCurrentTechAdvances(player));
+  }
+
+  static int sumIntegerMap(
+      final Function<TechAbilityAttachment, IntegerMap<UnitType>> mapper,
+      final UnitType ut,
+      final Collection<TechAdvance> techAdvances) {
+    return techAdvances.stream()
+        .map(TechAbilityAttachment::get)
+        .filter(Objects::nonNull)
+        .map(mapper)
+        .mapToInt(m -> m.getInt(ut))
+        .sum();
   }
 
-  private boolean getUnitAbilitiesGained(
+  private int getUnitAbilitiesGained(
```

Comment:
Why not keep it a boolean? Wouldn't it be nicer to just overload the `getCached` method instead?

Reply:
The map value type is Integer, so we'd either need two maps or have the type be Object and have getCached() cast it. WDYT?